pennylane/ftqc/pauli_tracker.py

```diff
@@ -235,3 +244,353 @@ def commute_clifford_op(clifford_op: Operator, xz: List[Tuple[int, int]]) -> Lis
+                )
+            by_op = _parse_rotz(ms)
+        elif isinstance(op, RotXZX):
+            if _t_ops_record[op.wires[0]] is None and idx < tape.num_wires:
+                _t_ops_record[op.wires[0]] = op
+            else:
+                raise NotImplementedError(
+                    "Current implement only support one non-Clifford gate per wire at the beginning of the circuit."
+                )
+            by_op = _parse_rotxzx(ms)
+        elif isinstance(op, _PAULIS):
+            continue
+        else:
+            raise NotImplementedError(f"{op.name} is not supported.")
+
+        by_ops.append(by_op)
+
+        mid_meas_offset += gate_offset
+
+    # To use list in a stack manner
+    by_ops.reverse()
+    return by_ops, ops
+
+
+def _get_xz_record(num_wires: int, by_ops: List[Tuple[int, int]], ops: List[Operator]):
+    """Commutate/merge the Pauli/byproduct ops of a Clifford circuit.
+
+    Args:
+        num_wires (int): Number of wires of the quantum state.
+        by_ops (list): List of byproduct operators for Clifford gates
+        ops (list): List of Clifford/Pauli/StatePrep operations.
+
+    Return:
+        The final recorded x and z for each wire.
+    """
+    x_record = math.zeros(num_wires, dtype=np.uint8)
+    z_record = math.zeros(num_wires, dtype=np.uint8)
+
+    for op in ops:
+        wires = list(op.wires)
+
+        # Get the recorded xz
+        xz = [(x_record[wire], z_record[wire]) for wire in wires]
+
+        # Updated xz
+        new_xz = []
+        if isinstance(op, _NON_CLIFFORD_GATES_SUPPORTED):
+            # Branch for non-Clifford gates
+            by_op = by_ops.pop()
+            new_xz.append(math.bitwise_xor(by_op[0], xz[0]))
+        elif isinstance(op, _CLIFFORD_GATES_SUPPORTED):
+            # Branch for Clifford gates
+            # Step 1: Commutate the recorded xz with the Clifford gate to a new xz.
+            xz_commutated = commute_clifford_op(op, xz)
+
+            # Step 2: Merge the new xz with the byproduct by_op
+            by_op = by_ops.pop()
+            for _by_op, _xz_comm in zip(by_op, xz_commutated):
+                new_xz.append(math.bitwise_xor(_by_op, _xz_comm))
+        else:  # branch for Paulis
+            # Commutate step is skipped.
+            # Get the new xz by merging the recorded xz with the Pauli ops directly.
+            new_xz.append(math.bitwise_xor(pauli_to_xz(op), xz[0]))
+        # Assign the updated the xz to the x, z record
+        for idx, wire in enumerate(wires):
+            x_record[wire], z_record[wire] = new_xz[idx]
+
+    return x_record, z_record
+
+
+def _apply_measurement_correction_rule(x: np.uint8):
+    """Get the phase correction factor based on the recorded `x` an `z` of the target wire and the corresponding
+    observable.
+
+        Args:
+            x (np.uint8): Recorded x at the target wire of ob.
+
+        Return:
+            Phase correction factor.
+    """
+    return -1 if x == 1 else 1
```

Comment:
So this is the table of:
```markdown
ip | out
---
000 | 1
001 | -1
010 | 1
011 | 1
100 | 1
101 | 1
110 | 1
111 | 1
```

or do we expect the input to be strictly binary (ie just `000` and `001`)?
If so, we can do `(-1)**x`, which is 1 and -1 respectively.